engine/execution/checker/engine.go

```diff
@@ -0,0 +1,109 @@
+package checker
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/consensus/hotstuff/notifications"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+type Engine struct {
+	notifications.NoopConsumer // satisfy the FinalizationConsumer interface
+
+	unit      *engine.Unit
+	log       zerolog.Logger
+	state     protocol.State
+	execState state.ExecutionState
+	sealsDB   storage.Seals
+}
+
+func New(
+	logger zerolog.Logger,
+	state protocol.State,
+	execState state.ExecutionState,
+	sealsDB storage.Seals,
+) *Engine {
+	return &Engine{
+		unit:      engine.NewUnit(),
+		log:       logger.With().Str("engine", "checker").Logger(),
+		state:     state,
+		execState: execState,
+		sealsDB:   sealsDB,
+	}
+}
+
+func (e *Engine) Ready() <-chan struct{} {
+	// make sure we will run into a crashloop if result gets inconsistent
+	// with sealed result.
+
+	finalized, err := e.state.Final().Head()
+
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("could not get finalized block on startup")
+	}
+
+	err = e.checkLastSealed(finalized.ID())
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("execution consistency check failed on startup")
+	}
+	return e.unit.Ready()
+}
+
+func (e *Engine) Done() <-chan struct{} {
+	return e.unit.Done()
+}
+
+// when a block is finalized check if the last sealed has been executed,
+// if it has been executed, check whether if the sealed result is consistent
+// with the executed result
+func (e *Engine) OnFinalizedBlock(block *model.Block) {
+	err := e.checkLastSealed(block.BlockID)
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("execution consistency check failed")
+	}
+}
+
+func (e *Engine) checkLastSealed(finalizedID flow.Identifier) error {
+	// TODO: better to query seals from protocol state,
+	// switch to state.Final().LastSealed() when available
+	seal, err := e.sealsDB.ByBlockID(finalizedID)
+	if err != nil {
+		return fmt.Errorf("could not get the last sealed for the finalized block: %w", err)
+	}
```

Comment:
The previous approach does not work. I thought `state.Sealed().Commit()` gives me the statecommit of the sealed block, but actually is the statecommit of the _last sealed_ included in the sealed block. No wonder the tests was failing

The `Commit` method is confusing, I would suggest to remove it, we could instead let it return a seal. Or replace this method with `LastSealed() (*flow.Seal, error)`, which I remembered has been implemented along with Epoch, right? @jordanschalm 

I checked no one is using the `Commit` method, except tests.

Reply:
Yes we added `LastSealedResult` in the epochs branch. `Commit` is still there, but I agree it's confusing.